### DISCUSSION TITLE
conditional Variables

### DISCUSSION THREAD

**Discussion by cpt-n3mo** [graphql/graphql-js#3425] [Q&A] [answered]
Good day,

im pritty new to graphql, it seems nice, but im having some trouble with the following..
i made a query like below:

```
query myProducts($catagory: String!, $make: String!, $subCatagory: [String!]!){
    myProducts(filter:{
      catagory:{ name:{ _eq: $catagory}}
      make:{ name: {_eq: $make }}
      subCatagory:{ mySubCatagories_id: {name: {_in: $subCatagory }}}
      }){
      name
      image{
        id
      }
      smallText
    }
  }
```
now this query will only run successfully when all the variables are set, but as im using it as a filter it would be nice if it runs even though not all variables are set.. how do i do this in graphql.. ??? is this possible.. ? Documentation is very confusing..

**Comment by saihaj** [comment, marked as answer]
Whenever you use `!` that type is a `non-nullable`/`required` type. Looks like from your query that all your filters are `non-nullable` arguments.
